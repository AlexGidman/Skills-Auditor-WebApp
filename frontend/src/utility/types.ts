export const systemRoleS = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = systemRoleS[0];
export const MANAGER_SR = systemRoleS[1];
export const ADMIN = systemRoleS[2];

export const jobRoleS = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = jobRoleS[0];
export const SENIOR_DEVELOPER = jobRoleS[1];
export const MANAGER_JR = jobRoleS[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const skillLevelS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = skillLevelS[0];
export const BASIC = skillLevelS[1];
export const INTERMEDIATE = skillLevelS[2];
export const HIGH = skillLevelS[3];
export const EXPERT = skillLevelS[4];

export type Category = {
    id: string;
    name: string;
};

export type User = {
    id?: string;
    firstName: string;
    lastName: string;
    email: string;
    password: string;
    jobRole: string; // TODO change this when type changes
    systemRole: string; // TODO change this when type changes
};

export type DirectReport = {
    id: string;
    report: User;
};

export type Skill = {
    id: string;
    name: string;
    categoryId: string;
    category: Category;
};

export type StaffSkill = {
    id: string;
    userId: string;
    skillId: string;
    skill: Skill;
    skillLevel: string;
    notes: string;
    expiryDate: string;
};
