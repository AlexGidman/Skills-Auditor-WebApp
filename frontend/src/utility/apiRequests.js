import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
