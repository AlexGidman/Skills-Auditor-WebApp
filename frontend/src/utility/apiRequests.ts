import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
