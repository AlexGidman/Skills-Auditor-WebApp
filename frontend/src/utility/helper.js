import { useEffect, useState } from "react";
import { useOutletContext, useNavigate } from "react-router-dom";
import {
    ADMIN,
    BASIC,
    EXPERT,
    HIGH,
    NONE,
    INTERMEDIATE,
    MANAGER_SR,
    SYSTEM_ROLES,
    TOKEN_COOKIE,
} from "./types";
import { useCookies } from "react-cookie";

export const useLogout = () => {
    const [, , removeCookie] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();
    const [logout, setLogout] = useState(false);

    useEffect(() => {
        if (logout) {
            setLogout(false);
            removeCookie(TOKEN_COOKIE);
            navigate("/login");
        }
    }, [logout, navigate, removeCookie]);

    return () => setLogout(true);
};

export const hyphenate = (string) => {
    return string.trim().replace(" ", "-");
};

export const useAPI = (apiCallback, varArray = [], lazy = false, auth = true) => {
    const [cookies] = useCookies(TOKEN_COOKIE);
    const navigate = useNavigate();

    const [callAPI, setCallAPI] = useState(!lazy);
    const [data, setData] = useState(null);
    const [loading, setLoading] = useState(!lazy);
    const [error, setError] = useState(null);

    const callback = () => {
        setLoading(true);
        setCallAPI(true);
    };

    const authHeaders = {
        headers: {
            authorization: cookies[TOKEN_COOKIE],
        },
    };

    useEffect(() => {
        if (callAPI) {
            apiCallback
                .apply(null, auth ? [authHeaders, ...varArray] : [...varArray])
                .then((response) => {
                    setData(response.data);
                })
                .catch((error) => {
                    if (unauthorisedError(error)) {
                        navigate("/login");
                    } else {
                        setError(error.response.data.error);
                    }
                })
                .finally(() => {
                    setLoading(false);
                    setCallAPI(false);
                });
        }
    }, [apiCallback, auth, callAPI, navigate, varArray]);

    return { data, loading, error, callback };
};

export const getSelectOptionsFromArray = (valArray = [], labelArray = []) => {
    return valArray.length === labelArray.length
        ? valArray.map((val, i) => {
              return { value: val, label: labelArray[i] };
          })
        : valArray.map((val) => {
              return { value: val, label: val };
          });
};

export const getAssignableSystemRoleOptions = (currentUserSystemRole) => {
    const systemRolesOptions = getSelectOptionsFromArray(SYSTEM_ROLES);
    switch (currentUserSystemRole) {
        case ADMIN:
            return systemRolesOptions;
        case MANAGER_SR:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option;
            });
        default:
            return systemRolesOptions.filter((option) => {
                return option.label !== ADMIN && option.label !== MANAGER_SR && option;
            });
    }
};

export const isValidString = (val, min = 0, max = 100) => {
    return !!val && val.length >= min && val.length <= max;
};

export const isValidEmail = (val) => {
    return /\S+@\S+\.\S+/.test(val);
};

export const isValidUser = (user) => {
    return (
        !!user.firstName &&
        isValidString(user.firstName, 1) &&
        !!user.lastName &&
        isValidString(user.lastName, 1) &&
        !!user.email &&
        isValidEmail(user.email) &&
        (!!user.password ? isValidString(user.password, 10, 16) : true)
    );
};

export const useIsAdminOrManager = () => {
    const [currentUser, setShowToast] = useOutletContext();
    const navigate = useNavigate();

    useEffect(() => {
        if (
            currentUser &&
            currentUser.system_role !== ADMIN &&
            currentUser.system_role !== MANAGER_SR
        ) {
            setShowToast({ error: "Unauthorised route" });
            navigate("/");
        }
    });
};

const unauthorisedError = (error) => error.response.status === 401 || error.response.status === 403;

export const formatDate = (_date) => {
    if (!_date) return "N/A";
    let newDate = new Date(_date);
    newDate.toLocaleDateString();
    return newDate.toLocaleDateString();
};

export const formatSkillLevel = (_skillLevel) => {
    if (_skillLevel === 5) {
        return EXPERT;
    } else if (_skillLevel === 4) {
        return HIGH;
    } else if (_skillLevel === 3) {
        return INTERMEDIATE;
    } else if (_skillLevel === 2) {
        return BASIC;
    } else {
        return NONE;
    }
};

export const convertSkillLevelName = (_skillLevel) => {
    if (_skillLevel === "Expert") {
        return 5;
    } else if (_skillLevel === "High") {
        return 4;
    } else if (_skillLevel === "Intermediate") {
        return 3;
    } else if (_skillLevel === "Basic") {
        return 2;
    } else {
        return 1;
    }
};

export const expiryDateNullFormatter = (_expiryDate) => {
    if (_expiryDate === null) {
        return "N/A";
    } else {
        return _expiryDate;
    }
};
