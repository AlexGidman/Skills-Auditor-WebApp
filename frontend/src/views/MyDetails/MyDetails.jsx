import React from "react";
import { useOutletContext } from "react-router-dom";
import { LinkButton } from "../../components/LinkButton/LinkButton";

import styles from "./MyDetails.module.css";

export const MyDetails = () => {
    const [currentUser] = useOutletContext();

    return (
        <>
            <header>
                <h1>My Details</h1>
            </header>
            <section className={styles.section}>
                <table className={styles.table}>
                    <tbody>
                        <tr>
                            <th>First name</th>
                            <td>{currentUser.forename}</td>
                        </tr>
                        <tr>
                            <th>Last name</th>
                            <td>{currentUser.surname}</td>
                        </tr>
                        <tr>
                            <th>Email</th>
                            <td>{currentUser.email}</td>
                        </tr>
                        <tr>
                            <th>Job role</th>
                            <td>{currentUser.job_role}</td>
                        </tr>
                        <tr>
                            <th>System role</th>
                            <td>{currentUser.system_role}</td>
                        </tr>
                    </tbody>
                </table>
                <LinkButton
                    className={styles.button}
                    name="Edit user details"
                    path={`/edit/user/${currentUser.id}`}
                />
            </section>
        </>
    );
};
