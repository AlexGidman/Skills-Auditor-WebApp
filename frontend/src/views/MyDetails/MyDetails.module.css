.button {
    margin-top: var(--s2);
    margin-left: auto;
    margin-right: auto;
}

.section {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    margin: 0 20%;
}

@media (max-width: 900px) {
    .section {
        margin: 0;
    }
    .button {
        margin-left: unset;
        margin-right: unset;
    }
}
