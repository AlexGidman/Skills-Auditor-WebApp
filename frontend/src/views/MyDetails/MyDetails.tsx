import React from "react";
import { useOutletContext } from "react-router-dom";
import { LinkButton } from "../../components/LinkButton/LinkButton";

import styles from "./MyDetails.module.css";

export const MyDetails = () => {
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser] = useOutletContext();

    return (
        <>
            <header>
                <h1>My Details</h1>
            </header>
            <section className={styles.section}>
                <table className={styles.table}>
                    <tbody>
                        <tr>
                            <th>First name</th>
                            <td>{currentUser.firstName}</td>
                        </tr>
                        <tr>
                            <th>Last name</th>
                            <td>{currentUser.lastName}</td>
                        </tr>
                        <tr>
                            <th>Email</th>
                            <td>{currentUser.email}</td>
                        </tr>
                        <tr>
                            <th>Job role</th>
                            <td>{currentUser.jobRole}</td>
                        </tr>
                        <tr>
                            <th>System role</th>
                            <td>{currentUser.systemRole}</td>
                        </tr>
                    </tbody>
                </table>
                <LinkButton
                    className={styles.button}
                    name="Edit user details"
                    path={`/edit/user/${currentUser.id}`}
                />
            </section>
        </>
    );
};
