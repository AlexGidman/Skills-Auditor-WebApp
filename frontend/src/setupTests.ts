// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
