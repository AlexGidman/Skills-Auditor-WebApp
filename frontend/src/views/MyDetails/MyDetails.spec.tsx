import { render, screen } from "@testing-library/react";
import { MyDetails } from "./MyDetails";
import { mockApiRequests } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import React from "react";
import { ADMIN, MANAGER_JR, User } from "../../utility/types";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

const mockUser: User = {
    id: "1",
    firstName: "John",
    lastName: "Smith",
    email: "John@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

describe("MyDetails", () => {
    it("renders correctly with currentUser's details", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({ currentUser: mockUser });
        // @ts-ignore TODO: fix this
        mockApiRequests.getUserDetails.mockImplementation(() =>
            Promise.resolve({ data: mockUser }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <MyDetails />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "My Details" })).toBeInTheDocument();

        expect(screen.getByRole("table")).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "First name" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser.firstName })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Last name" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser.lastName })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Email" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser.email })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "Job role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser.jobRole })).toBeInTheDocument();
        expect(screen.getByRole("columnheader", { name: "System role" })).toBeInTheDocument();
        expect(screen.getByRole("cell", { name: mockUser.systemRole })).toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit user details" })).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should have LinkButton that links to /edit/user/:userId when data successful", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({ currentUser: mockUser });
        // @ts-ignore TODO: fix this
        mockApiRequests.getUserDetails.mockImplementation(() =>
            Promise.resolve({ data: mockUser }),
        );
        render(
            <ReactDom.MemoryRouter>
                <MyDetails />
            </ReactDom.MemoryRouter>,
        );
        const editUserDetailsLinkButton = await screen.findByRole("link", {
            name: "Edit user details",
        });
        expect(editUserDetailsLinkButton).toHaveAttribute("href", `/edit/user/${mockUser.id}`);
    });
});
