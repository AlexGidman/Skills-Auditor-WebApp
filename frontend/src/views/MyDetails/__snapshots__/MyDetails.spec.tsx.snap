// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyDetails renders correctly with currentUser's details 1`] = `
<div>
  <header>
    <h1>
      My Details
    </h1>
  </header>
  <section
    class="section"
  >
    <table
      class="table"
    >
      <tbody>
        <tr>
          <th>
            First name
          </th>
          <td>
            John
          </td>
        </tr>
        <tr>
          <th>
            Last name
          </th>
          <td>
            Smith
          </td>
        </tr>
        <tr>
          <th>
            Email
          </th>
          <td>
            John@email.com
          </td>
        </tr>
        <tr>
          <th>
            Job role
          </th>
          <td>
            Manager
          </td>
        </tr>
        <tr>
          <th>
            System role
          </th>
          <td>
            Admin
          </td>
        </tr>
      </tbody>
    </table>
    <a
      class="component button"
      href="/edit/user/1"
    >
      Edit user details
    </a>
  </section>
</div>
`;
